lib/pbench/agent/redis.py

```diff
@@ -222,39 +222,77 @@ def emit(self, record):
+def wait_for_conn_and_key(redis_server: redis.Redis, key: str, prog: str) -> str:
     """wait_for_conn_and_key - convenience method of both the Tool Meister and
     the Tool Data Sink to startup and wait for an initial connection to the
     Redis server, and for the expected key to show up.
+
+    Arguments:
+
+        redis_server: a Redis client object
+        key: the key name as a string
+        prog: the program name of our caller
+
+    Returns the payload value of the key as a string.
+
+    If successful on the first connection attempt, and the value of the key is
+    retrieved, no messages are issued on stdout.  If the key does not exist yet,
+    we'll report the missing key every 10 seconds from the last successful
+    connection.
+
+    If on the first connection attempt the connection to the Redis server fails,
+    we emit a message to that effect, once.  After that, we'll report each time
+    we are disconnected or re-connected to the Redis server.
     """
     # Loop waiting for the key to show up.
     connected = None
     payload = None
+    attempts = 0
+    errors = 0
+    missing = 0
     while payload is None:
+        attempts += 1
         try:
             payload = redis_server.get(key)
         except redis.ConnectionError:
+            errors += 1
+            # Reset the missing count if we get disconnected
+            missing = 0
             if connected is None:
+                # Only emit this message once for the initial attempt.
                 print(
-                    f"{prog}: waiting to connect to redis server {redis_host}:{redis_port}",
+                    f"{prog}: waiting to connect to Redis server {redis_server}",
                     flush=True,
                 )
                 connected = False
             elif connected:
+                # We always report disconnections.
                 print(
-                    f"{prog}: disconnected from redis server {redis_host}:{redis_port}",
+                    f"{prog}: disconnected from Redis server {redis_server}",
                     flush=True,
                 )
                 connected = False
             time.sleep(1)
         else:
-            if not connected:
+            if connected is False:
+                # We always report re-connections (connections was not None).
                 print(
-                    f"{prog}: connected to redis server {redis_host}:{redis_port}",
-                    flush=True,
+                    f"{prog}: connected to Redis server {redis_server}", flush=True,
                 )
-                connected = True
+            connected = True
             if payload is None:
-                print(f'{prog}: key, "{key}" does not exist yet', flush=True)
+                missing += 1
+                if (missing % 10) == 0:
+                    # Only emit missing key notice every 10 seconds
+                    print(f"{prog}: key '{key}' does not exist yet", flush=True)
```

Comment:
I can see periodic messages if we remain disconnected for a period of time, but this could get fairly noisy especially during startup deployment. Maybe once a minute (or even more) rather than every 10 seconds?

Reply:
> Why do periodic prints when we're waiting only if we're connected? Isn't a "heartbeat" worthwhile even if we're disconnected?

> I can see periodic messages if we remain disconnected for a period of time, but this could get fairly noisy especially during startup deployment. Maybe once a minute (or even more) rather than every 10 seconds?

This has to do with what I found for behavior when trying to orchestrate the individual components for the Tool Meister sub-system.

When `pbench-tool-meister-start` is responsible for the orchestration, it very carefully and deliberately starts things up in a certain order: Redis server first (confirming it is there and working), Tool Data Sink second (confirming it is there and listening), and then all the Tool Meisters.

In this scenario, the scheme we are following for `wait_for_conn_and_key()` will very likely succeed on the first attempt.  So the complexity here is not due to that case.

When something else is responsible for starting things up, we can have any number of scenarios, e.g. TMs first, TDS 2nd, Redis server last; TDS first, Redis server 2nd, TMs last; 1 TM first, Redis server next, TDS next, etc.

However, the subtlety is that something has to load the keys into the Redis server before the TDS and TMs can begin their operation.  So losing a connection to the Redis server before the TDS and TMs receive their marching orders is in consequential.  As long as it comes back, and they eventually find their keys, we're good.

What we can't tolerate (yet) is getting our marching orders and then having the Redis server disappear.

This sequence allowed me to bring a Redis server up by hand, add keys slowly, and in different orders so that when I screwed up I could just reset the Redis server before it all got going.  And the messages let me know that once I have a Redis server up and running, I had better populate it.